staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -667,8 +669,39 @@ func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 	c.startDispatching(event)
 
 	// Since add() can block, we explicitly add when cacher is unlocked.
+
+	// Dispatching event in nonblocking way first, which make faster watchers
+	// not be blocked by slower ones.
+	c.blockedWatchers = c.blockedWatchers[:0]
 	for _, watcher := range c.watchersBuffer {
-		watcher.add(event, c.timer, c.dispatchTimeoutBudget)
+		if !watcher.nonblockingAdd(event) {
+			c.blockedWatchers = append(c.blockedWatchers, watcher)
+		}
+	}
+
+	if len(c.blockedWatchers) > 0 {
+		// dispatchEvent is called very often, so arrange
+		// to reuse timers instead of constantly allocating.
+		startTime := time.Now()
+		timeout := c.dispatchTimeoutBudget.takeAvailable()
+		c.timer.Reset(timeout)
+
+		// Make sure every watcher will try to send event without blocking first,
+		// even if the timer has already expired.
+		timer := c.timer
+		for _, watcher := range c.blockedWatchers {
+			if !watcher.add(event, timer) {
+				// No time left, clean the timer by set it to nil.
+				timer = nil
+			}
+		}
+		if !c.timer.Stop() {
+			// Consume triggered (but not yet received) timer event
+			// so that future reuse does not get a spurious timeout.
+			<-c.timer.C
```

Comment:
I think this is incorrect. https://golang.org/pkg/time/#Timer.Stop says that this pattern can be used to drain the timer ONLY if we know that the program has not received from c.timer.C.

It's not the case here (it was before the change) -- in case of timeout, watcher.add will consume c.timer.C and return false. Then we will try to call timer.Stop (it will return false as the timer has already expired) and here we will try to read from the channel again, hanging forever.

Reply:
I think it's worth to add a test case for timeout scenario. It will be a way easier to find regressions like that instead of debugging 5000 machines test.

Reply:
sorry for the mistack, I will fix it with a test.